plum/parametric.py

```diff
@@ -54,16 +56,33 @@ def __call__(cls, *args, **kw_args):
         # and then call the equivalent of `T(arg1, arg2, **kw_args)`.
 
         if not cls.is_concrete:
-            type_parameter = tuple(type(arg) for arg in args)
-            if len(type_parameter) == 1:
-                type_parameter = type_parameter[0]
+            type_parameter = cls.__infer_type_parameter__(*args, **kw_args)
             T = cls[type_parameter]
         else:
             T = cls
 
         # Calls `__new__` and `__init__`.
         return type.__call__(T, *args, **kw_args)
 
+    def __infer_type_parameter__(cls, *args, **kw_args):
+        """Function called when the constructor of this parametric type is called
+        before the parameters have been specified.
+
+        The default behaviour is to take as parameters the type of every argument,
+        but this behaviour can be overridden by redefining this function on the
+        metaclass.
+
+        Args:
+            *args: the argument-values passed to the __init__ method.
+
+        Returns:
+            A type or tuple of types.
+        """
+        type_parameter = tuple(type(arg) for arg in args)
```

Comment:
Should we consider using `type_of` here?

Reply:
fine by me